WiimoteReal: Don't block on refresh

leoetlino opened the pull request:
This changes Refresh() to use the existing scanning thread to scan for devices, instead of running the scan on the UI thread and blocking it. (Should fix [issue 8992](https://dolp.in/i8992))

This also makes the UI thread not block when Continuous Scanning is enabled then disabled, and removes a random 0-500ms delay on exit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3988)
<!-- Reviewable:end -->

JMC47 commented:
Very, very nice.

delroth commented on Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp:
```diff
@@ -471,10 +478,15 @@ void WiimoteScanner::ThreadFunc()
 
   while (m_run_thread.load())
   {
+    m_scan_event.WaitFor(std::chrono::milliseconds(500));
```
Couldn't that use a Common::Event instead of a polling loop?

delroth replied:
Why a timeout here?

delroth commented on Source/Core/Core/HW/WiimoteReal/WiimoteReal.cpp:
```diff
@@ -440,18 +440,25 @@ void WiimoteScanner::WantBB(bool do_want)
+void WiimoteScanner::StartScanning(bool scan_only_once)
 {
+  m_should_scan.store(true);
+  m_should_scan_only_once.store(scan_only_once);
```
This should be set before triggering the WiimoteScanner thread to avoid getting wrong the value.

delroth commented:
As I was just saying on IRC: it feels like the logic here is being overly complex and that it could be simplified much further. For example, if we're always going to do scanning in a thread, we can potentially get rid of a tons of things that were assuming scanning could happen in two different threads. We can also keep the thread running continuously instead of having StopScanning with two behaviours (stop / do not stop thread). etc.

leoetlino commented:
I have switched to enums (instead of booleans), and removed code that shouldn't be necessary anymore since scanning now only happens on one thread. Also made it clear that the scanning thread is always running.

delroth commented:
I'm not convinced about the thread safety of the new code. For example:

```
void InterruptChannel(int _WiimoteNumber, u16 _channelID, const void* _pData, u32 _Size);
void ControlChannel(int _WiimoteNumber, u16 _channelID, const void* _pData, u32 _Size);
```

These two functions now perform unsynchronized accesses to g_wiimotes but nothing guarantees that they would be called from the scanning thread. In fact, g_wiimotes itself is global, which is a fairly terrible design...

Can you explain why removing the lock is safe? I understand that it was recursive before because of the potential "double scan" issue, and that's probably useless now, but still feels like it should be a simple mutex.

leoetlino commented:
You're right; that was not very safe. I have added back a simple mutex in places where they are needed.

degasus commented:
Do you have an idea about android? https://forums.dolphin-emu.org/Thread-wiimote-dolphinbar-not-refreshed-after-5-0-151

leoetlino commented:
Huh, that's a weird regression… So basically stopping a game prevents it from working again?

My first guess would be that it relies on WiimoteReal::Refresh() (which is basically now a noop if continuous scanning is enabled) and for some reason it needs the scanning thread to really be stopped and then restarted. I can't test it on Android (no ARMv8 device), but I'll try to see if there is something I can do…
